### DISCUSSION TITLE
Strange compiling error with arduino-pico and platformio

### DISCUSSION THREAD

**Discussion by schinsve** [earlephilhower/arduino-pico#2224] [Q&A]
Hey guys, 

I am encountering a strange error that I was not able to find a solution to anywhere online and Im starting to think, that something between arduino-pico and platformio is going wrong. It might be related to my insufficient C/C++ knowledge as well. I'm more profficient with Python, but for my bachelors thesis I had to change the language. Help would be much appreciated.

Now to the error itself. I have cropped the parts that went well. If you need those, I can provide them off course:

```
.
.
.
Compiling .pio\build\pico\FrameworkArduino\wiring_shift.cpp.o
Archiving .pio\build\pico\libFrameworkArduino.a
Indexing .pio\build\pico\libFrameworkArduino.a
Linking .pio\build\pico\firmware.elf
c:/users/ssven/.platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/12.3.0/../../../../arm-none-eabi/bin/ld.exe: .pio\build\pico\src\main.cpp.o: in function `main':
main.cpp:(.text.startup.main+0x4): undefined reference to `stdio_init_all'
c:/users/ssven/.platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/12.3.0/../../../../arm-none-eabi/bin/ld.exe: .pio\build\pico\lib2a5\libpico-tidbits.a(uart_io.c.o): in function `uart_io_start':
uart_io.c:(.text.uart_io_start+0x26): undefined reference to `getchar_timeout_us'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\pico\firmware.elf] Error 1
```

If I correctly understand the error, the linker is not able to find the definitions of stdio_init_all and getchar_timeout_us, which is strange, since the docs (https://arduino-pico.readthedocs.io/en/latest/sdk.html) mention, that the pico-sdk is included with the arduino core. I dont get any errors in my main.cpp too. The definitions are available there. Only the linker cant find them.

Solutions I have tried, from other forums were:
- Including the libraries via: extern "C" {}
- Enabeling long paths on Windows and for git
- Adding the pico-sdk via lib_deps
- Using different build_flags (Might have done that incorrectly though)

For reference purposes Ill copy my platformio.ini here:

```
[env:pico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
framework = arduino
board_build.core = earlephilhower
lib_deps = 
	https://github.com/apparanto/pico-tidbits.git
```

And here a test program that is generating the error as well:

```
/**
 * @file main.c
 * @author Robert Portier
 * @brief 
 * @version 0.1
 * @date 2022-02-13
 * 
 * @copyright Copyright (c) 2022
 * 
 */

extern "C" {
    #include <time.h>
    #include <pico/stdio.h>
    #include "sim7080g/sim7080g_cmd.h"

}



int main() {

    stdio_init_all();
    sleep_ms(3000);
    printf("Initializing. Please wait...\n");

    time_t t = time(NULL);
    struct tm tm = *localtime(&t);
    printf("%d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour, tm.tm_min, tm.tm_sec);
    
    if (sim7080g_init(0)) {
        printf("Initialization succesful\n");
    } else {
        printf("Initialization failed\n");
    }

    sim7080g_command();

    while (true) {
        tight_loop_contents();
    }
    return 0;
}
```
If you need any further information I am happy to provide it, if I can
Thank you very much for your help

**Comment by earlephilhower** [comment]
Those APIs are not valid in Arduino-land.  Use `Serial.printf` or `Serial1/2.printf` for output.  The USB and UARTS are owned by the core and those objects.

**Reply by schinsve** [reply]
Thank you very much. That helped a huge bit. I wish you a wonderful weekend :)
